### DISCUSSION TITLE
RegEx rule system

### DISCUSSION THREAD

**Discussion by aparatext** [ThioJoe/YT-Spammer-Purge#47] [Ideas]
Allow a user to specify a local text file or an online URL with RegEx rules for blacklisting channel names, channel IDs, and comments as an argument.

Here's an example of a rule list.
```
// No arrow-up emoji in the channel name
channel_name: ⬆|🔼|🔺

// No comments with Bitcoin wallets in channel name or comment.
channel_name, comment: [13][a-km-zA-HJ-NP-Z1-9]{25,34}$

// No comments from ThioJoe
channel_id: ^UCQSpnDG3YsFNf5-qHocF-WQ
```

**Comment by ArjixWasTaken** [comment]
I don't see how using regex to filter channel IDs is useful...
But using a list of channel IDs instead of a single channel ID does sound better.

**Comment by aparatext** [comment]
RegEx can allow for more granular filtering. A rule set can contain RegEx to filter phone numbers, links, common obfuscation techniques, and more. As for filtering channel IDs, it might work as a replacement for channel list filtering.

**Comment by ArjixWasTaken** [comment]
If you reread my message, I only talked about the uselessness of regex for channel IDs (I am saying this since you replied about all the things you said when I only talked about one of them)
Since it would be more clean and make more sense to use a list instead.

I did not disagree or doubt that regex is useful for the other stuff.
Just the channel IDs

Whilst using regex for the channel IDs makes it possible to have a one-liner, it would also make it look ugly.

Besides, if it's about one-liners we can use a list comprehension instead.

```python
channel_ids = ["xx", "yy"]

test_input = "xx"

is_bot: bool = any(x==test_input for x in channel_ids)
```

I do love using regex, but in some cases it's not "pythonic" to use it, if you get what I mean.

The main reason why I am against regex for the channel IDs is because:
- It would be a huge string at some point (if you want to add a lot of channel IDs)
- Regex is for pattern matching, and what you are suggesting does not contain patterns (again, I'm talking about the channel IDs only).

**Comment by http403** [comment]
How should the list be implemented? A json or yaml file, or a python config script? Former one is easier to edit (in term of syntax) and less coupled with Python.

**Reply by ArjixWasTaken** [reply]
we already have a config file for the API key, we can just put the list there

**Reply by http403** [reply]
I'm a bit reluctant to combine rule sets with config as I can see in the future the rule set will be quite big to fit into one file. I'm more prefer in using a rules directory with multiple files.

**Reply by rbeer** [reply]
You make a list. channel_ids.txt. No need for JSON or any of that. One channel id per line and you're done.

1. Read the file
2. Split by `\n`
3. Use the list

**Reply by http403** [reply]
The RegEx system isn't for channel ids. It's more focus on content pattern matching, for example Bitcoin address, emails, and URLs. Of course you can have simple rule per line file too but that would mean you can't have conditions like rules only applies on the username.

**Comment by http403** [comment]
In addition to RegEx, I think a hybrid system that consist string matching and RegEx is better. Thoughts?

**Reply by http403** [reply]
It will be a combination with RegEx with the idea of #39 but in a file.
